File: 2018/src/day02.rs
use std::fs;

pub fn day02() {
    let input = "data/input-02.txt";
    let file = fs::read_to_string(input).unwrap();
    let vals: Vec<_> = file.lines().collect();

    let (mut num2, mut num3) = (0, 0);
    for val in &vals {
        let counts: Vec<_> = val.chars().map(|z| val.matches(z).count()).collect();
        if counts.contains(&2) {
            num2 += 1;
        }
        if counts.contains(&3) {
            num3 += 1;
        }
    }
    let res = num2 * num3;
    println!("Part 1: {res}");
    assert_eq!(res, 6200);

    let mut res = String::new();
    'outer: for (idx, val1) in vals.iter().enumerate() {
        for val2 in &vals[(idx + 1)..] {
            res = val1
                .chars()
                .zip(val2.chars())
                .filter_map(|(x, y)| if x == y { Some(x) } else { None })
                .collect();
            if val1.len() - res.len() == 1 {
                break 'outer;
            }
        }
    }
    println!("Part 2: {res}");
    assert_eq!(res, "xpysnnkqrbuhefmcajodplyzw");
}

File: 2018/julia/day02.jl
using DelimitedFiles

vals = readdlm("data/input-02.txt")[:, 1]
letters = unique([only(v) for val in vals for v in split(val, "")])

num2, num3 = 0, 0
for val in vals
    counts = [count(letter -> (letter == l), val) for l in letters]
    global num2 += 2 in counts
    global num3 += 3 in counts
end
res = num2 * num3
println("Part 1: $(res)")
@assert(res == 6200)

left = copy(vals)
for v1 in vals
    deleteat!(left, findall(x -> x == v1, left))
    for v2 in left
        if sum([i != j for (i, j) in zip(v1, v2)]) == 1
            global res = join([i for (i, j) in zip(v1, v2) if i == j])
        end
    end
end
println("Part 2: $(res)")
@assert(res == "xpysnnkqrbuhefmcajodplyzw")
